### PR TITLE
fix: Junk/NotJunk flags

### PR DESCRIPTION
**Pull request by SebastianKrupinski**
### Issue

Junk and NotJunk flags where not being set due to a invalid comparison. UI was sending "$junk" but it was being compared to "junk".
 
Screenshot 2024-11-05 181610

Allowed flags was also missing "notjunk"

image

### CODE REVIEW

**Comment by ChristophWurst** [comment]
How do we test this?

Found https://github.com/nextcloud/mail/pull/5370 to change ``junk`` to ``$junk``

File: src/store/actions.js [hunk 1]
```diff
@@ -998,8 +998,8 @@ export default {
 
 			try {
 				await setEnvelopeFlags(envelope.databaseId, {
-					$junk: !oldState,
-					$notjunk: oldState,
```
Comment (ChristophWurst) [inline src/store/actions.js:1002]: https://github.com/nextcloud/mail/blob/1d5e5b49c1485e9be70f22c4df9526805f5238a7/src/store/actions.js#L991, https://github.com/nextcloud/mail/blob/1d5e5b49c1485e9be70f22c4df9526805f5238a7/src/store/actions.js#L1011-L1021 and similar stay?

Comment (SebastianKrupinski) [reply src/store/actions.js:1002]: Yes, The UI handles the "$junk" but the backend compares the it as "junk", so in the UI they can all stay as "$junk" but need to have the $ sign stripped of when sent to the backend. The second image of allowed flags is comparison table, the first image is how they are compared/found.

**Comment by SebastianKrupinski** [comment]
> How do we test this?

The best way to test is to insert a break point in the OCA\Mail\Service\MailManager::filterFlags() function. That way you can see the values for the table and flag.

> Found #5370 to change `junk` to `$junk`

This correctly added the front end logic,

```
commit('flagEnvelope', {
	envelope,
	flag: '$notjunk',
	value: !oldState,
})
```

But incorrectly changed the back end logic,

```
setEnvelopeFlag(envelope.databaseId, '$junk', !oldState).catch((e) => {
```

File: src/store/actions.js (hunk 1 above)
Comment (SebastianKrupinski) [reply src/store/actions.js:1002]: Sure, Its actual the line you highlighted,

Before:
```
await setEnvelopeFlags(envelope.databaseId, {
	$junk: !oldState,
	$notjunk: oldState,
})
```
After:
```
await setEnvelopeFlags(envelope.databaseId, {
	junk: !oldState,
	notjunk: oldState,
})
```

**Comment by kesselb** [comment]
Imap Flags, my favorite topic ;)

Please don't spend too much time on this one. We are aware about a couple of issues with tags (e.g. https://github.com/nextcloud/mail/issues/10159), and it's on my to-do to work on it when I'm done with files_emailviewer and sieve. 

Your finding is correct, $junk is not part of the allowed flags and therefore the check will never return true. The list of allowed flags should only contain the system flags, which are mandatory from the RFC: https://www.ietf.org/rfc/rfc9051.html#name-flags-message-attribute. 

For system flags, we can skip the check if permflags are enabled and save a few bits. 

If Junk/NotJunk was not added on your system, I assume permflags are not enabled. Can you check that with the debugger?

**Comment by kesselb** [comment]
> Which brings me to the second logic error, the function then checks if '\ *' exists which is NOT an indicator for system flags or keywords, '\ *' 'indicates that custom flags/keywords (tags/labels) are supported.

:+1: 

We currently don’t handle cases where custom keywords are not allowed, but the keyword already exists.

Here’s what I suggest:

- Keep $junk as is.
- Remove non-system flags from allowed flags.
- Introduce a new method in MailManager to check if a given keyword exists or if we can create a new one.

Let me know your thoughts or if you have any suggestions on this approach!

**Comment by SebastianKrupinski** [comment]
> * Keep $junk as is.
> 
> * Remove non-system flags from allowed flags.
> 
> * Introduce a new method in MailManager to check if a given keyword exists or if we can create a new one.

Here is my thoughts,

1. We should change '$junk' in the UI and send it as 'junk', because all the other flags get sent without the $ or \ signs, this was the only one that was send to the back end differently. Otherwise we need to have all the flags sent with the sign from the UI.
2. I agree we show remove the keywords from the allowed flags. But we could also leave them as they are RFC defined and defined in horde (This would basically saves a request to the server for PERMANENTFLAGS check on every flag change, but there is no guarantee they are supported, although I would put my money that most systems support them).
3. This can be addressed pretty easily in the same function with a simple if statement, pull the supported flags then check for '\\ *' and if that fails check if the specific flag/keyword it supported. (I'll make these changes, it will be a quick modification)

I also have some thoughts on how we can make this way more efficient and work better. But I'll DM you with the idea, so that we don't pollute this chat.

**Comment by SebastianKrupinski** [comment]
@kesselb made the changes we discussed. The filterFlags method now handles all 3 cases, system flags, server defined/supported keywords and server custom keywords

**Review by kesselb — APPROVED** [review]

**Pull request merged**